File: pkg/@eser/cache/cache.ts
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

/**
 * Cache manager implementation.
 *
 * Provides versioned cache storage with XDG-compliant directory structure.
 *
 * @example
 * ```typescript
 * import { createCacheManager } from "@eser/cache";
 *
 * const cache = createCacheManager({
 *   app: { name: "my-cli", org: "eser" }
 * });
 *
 * // Get versioned cache path
 * const binaryPath = cache.getVersionedPath("1.0.0", "binary");
 *
 * // Ensure directory exists
 * await cache.ensureDir(runtime.path.dirname(binaryPath));
 *
 * // Check if cached
 * if (await cache.exists(binaryPath)) {
 *   console.log("Using cached binary");
 * }
 * ```
 *
 * @module
 */

import * as standardsRuntime from "@eser/standards/runtime";
import type {
  CacheEntry,
  CacheManager,
  CacheManagerOptions,
} from "./primitives.ts";
import * as xdg from "./xdg.ts";

/**
 * Creates a cache manager for the specified application.
 *
 * @param options - Cache manager configuration
 * @returns A CacheManager instance
 *
 * @example
 * ```typescript
 * const cache = createCacheManager({
 *   app: { name: "my-cli", org: "eser" }
 * });
 *
 * const path = cache.getVersionedPath("1.0.0", "binary");
 * // On Linux: ~/.cache/eser/my-cli/v1.0.0/binary
 * // On macOS: ~/Library/Caches/eser/my-cli/v1.0.0/binary
 * ```
 */
export const createCacheManager = (
  options: CacheManagerOptions,
): CacheManager => {
  const { app, baseDir } = options;

  const getCacheDir = (): string => {
    if (baseDir) {
      if (app.org) {
        return standardsRuntime.runtime.path.join(baseDir, app.org, app.name);
      }
      return standardsRuntime.runtime.path.join(baseDir, app.name);
    }
    return xdg.getAppCacheDir(app);
  };

  const resolvePath = (path: string): string => {
    if (standardsRuntime.runtime.path.isAbsolute(path)) {
      return path;
    }
    return standardsRuntime.runtime.path.join(getCacheDir(), path);
  };

  const getVersionedPath = (version: string, name: string): string => {
    if (baseDir) {
      // Normalize version to always have 'v' prefix
      const normalizedVersion = version.startsWith("v")
        ? version
        : `v${version}`;
      return standardsRuntime.runtime.path.join(
        getCacheDir(),
        normalizedVersion,
        name,
      );
    }
    return xdg.getVersionedCachePath(app, version, name);
  };

  const exists = async (path: string): Promise<boolean> => {
    const resolvedPath = resolvePath(path);
    return await standardsRuntime.runtime.fs.exists(resolvedPath);
  };

  const ensureDir = async (path: string): Promise<void> => {
    const resolvedPath = resolvePath(path);
    await standardsRuntime.runtime.fs.mkdir(resolvedPath, { recursive: true });
  };

  const list = async (): Promise<CacheEntry[]> => {
    const cacheDir = getCacheDir();
    const entries: CacheEntry[] = [];

    try {
      const dirExists = await standardsRuntime.runtime.fs.exists(cacheDir);
      if (!dirExists) {
        return entries;
      }

      for await (const entry of standardsRuntime.runtime.fs.readDir(cacheDir)) {
        const entryPath = standardsRuntime.runtime.path.join(
          cacheDir,
          entry.name,
        );

        let size = 0;
        let mtime: Date | null = null;

        try {
          const stat = await standardsRuntime.runtime.fs.stat(entryPath);
          size = stat.size;
          mtime = stat.mtime;
        } catch {
          // Ignore stat errors
        }

        entries.push({
          path: entry.name,
          name: entry.name,
          isDirectory: entry.isDirectory,
          size,
          mtime,
        });
      }
    } catch {
      // Return empty if directory doesn't exist or can't be read
    }

    return entries;
  };

  const remove = async (path: string): Promise<void> => {
    const resolvedPath = resolvePath(path);

    try {
      await standardsRuntime.runtime.fs.remove(resolvedPath, {
        recursive: true,
      });
    } catch {
      // Ignore if path doesn't exist
    }
  };

  const clear = async (): Promise<void> => {
    const cacheDir = getCacheDir();

    try {
      await standardsRuntime.runtime.fs.remove(cacheDir, { recursive: true });
    } catch {
      // Ignore if directory doesn't exist
    }
  };

  return {
    getCacheDir,
    getVersionedPath,
    exists,
    ensureDir,
    list,
    remove,
    clear,
  };
};

File: pkg/ajan/cachefx/cache.go
// Copyright 2023-present Eser Ozvataf and other contributors. All rights reserved. Apache-2.0 license.

package cachefx

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
)

// ErrNotFound is returned when a requested cache entry does not exist.
var ErrNotFound = errors.New("cache entry not found")

// AppIdentifier uniquely identifies an application for cache namespacing.
type AppIdentifier struct {
	// Org is an optional organisation or vendor prefix (e.g. "eser").
	Org string
	// Name is the required application name (e.g. "ajan").
	Name string
}

// Options configures a new Manager.
type Options struct {
	App AppIdentifier
	// BaseDir overrides the XDG cache home. Defaults to XdgCacheHome().
	BaseDir string
}

// Entry describes a single item in the cache directory.
type Entry struct {
	Path        string
	Name        string
	MtimeUnix   int64
	Size        int64
	IsDirectory bool
}

// Manager provides cache directory operations scoped to a single application.
type Manager struct {
	cacheDir string
}

// NewManager returns a Manager whose cache dir is rooted under BaseDir (or the
// XDG cache home when BaseDir is empty), namespaced by Org and Name.
func NewManager(opts Options) *Manager {
	base := opts.BaseDir
	if base == "" {
		base = XdgCacheHome()
	}

	var dir string
	if opts.App.Org != "" {
		dir = filepath.Join(base, opts.App.Org, opts.App.Name)
	} else {
		dir = filepath.Join(base, opts.App.Name)
	}

	return &Manager{cacheDir: dir}
}

// Dir returns the absolute path to the application cache directory.
func (m *Manager) Dir() string { return m.cacheDir }

// VersionedPath returns the path for a named artefact under a semver version.
// A leading "v" is normalised so "1.0.0" and "v1.0.0" resolve to the same path.
func (m *Manager) VersionedPath(version, name string) string {
	v := "v" + strings.TrimPrefix(version, "v")

	return filepath.Join(m.cacheDir, v, name)
}

// Exists reports whether path exists on disk.
func (m *Manager) Exists(path string) bool {
	_, err := os.Stat(path)

	return !errors.Is(err, fs.ErrNotExist)
}

// EnsureDir creates path and all parents if they do not already exist.
func (m *Manager) EnsureDir(path string) error {
	if err := os.MkdirAll(path, 0o755); err != nil { //nolint:gosec // directory creation
		return fmt.Errorf("ensure dir %q: %w", path, err)
	}

	return nil
}

// List returns all top-level entries inside the cache directory.
// Returns an empty slice (not an error) when the cache directory does not exist.
func (m *Manager) List() ([]Entry, error) {
	dirEntries, err := os.ReadDir(m.cacheDir)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return []Entry{}, nil
		}

		return nil, fmt.Errorf("list cache %q: %w", m.cacheDir, err)
	}

	result := make([]Entry, 0, len(dirEntries))

	for _, de := range dirEntries {
		info, err := de.Info()
		if err != nil {
			continue // skip entries whose stat fails (e.g. dangling symlinks)
		}

		result = append(result, Entry{
			Path:        filepath.Join(m.cacheDir, de.Name()),
			Name:        de.Name(),
			Size:        info.Size(),
			MtimeUnix:   info.ModTime().Unix(),
			IsDirectory: de.IsDir(),
		})
	}

	return result, nil
}

// Remove deletes the entry at path (file or directory tree).
func (m *Manager) Remove(path string) error {
	if err := os.RemoveAll(path); err != nil {
		return fmt.Errorf("remove cache entry %q: %w", path, err)
	}

	return nil
}

// Clear deletes the entire application cache directory and all its contents.
func (m *Manager) Clear() error {
	if err := os.RemoveAll(m.cacheDir); err != nil {
		return fmt.Errorf("clear cache %q: %w", m.cacheDir, err)
	}

	return nil
}
